Util/PagingIterator.php

```diff
@@ -54,41 +60,76 @@ public function __construct(Collection $collection)
             throw new NostoException('Page size not defined or not an integer');
         }
         $this->collection = $collection;
+        $this->lastPageNumber = $this->collection->getLastPageNumber();
     }
 
     /**
-     * Iterates through the collection in batches defined by the collection page size
-     * and applies closure to each item
-     * @param \Closure $closure
-     * @throws NostoException
+     * @inheritdoc
      */
-    public function each(\Closure $closure)
+    public function current()
     {
-        $curPage = 1;
-        $lastPage = $this->collection->getLastPageNumber();
-        do {
-            $this->collection->clear();
-            $this->collection->setCurPage($curPage);
-            $this->collection->each($closure);
-            ++$curPage;
-        } while ($curPage <= $lastPage);
+        return $this->collection;
     }
 
     /**
-     * Handles the pagination / batching and batching for a collection
-     *
-     * @param \Closure $closure
-     * @throws NostoException
+     * @inheritdoc
+     */
+    public function next()
+    {
+        ++$this->currentPageNumber;
+        $this->page($this->currentPageNumber);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function key()
+    {
+        return $this->collection->getCurPage();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function valid(): bool
+    {
+        if ($this->currentPageNumber <= $this->lastPageNumber) {
+            return true;
+        }
+        return false;
```

Comment:
Can be written as:

        return $this->currentPageNumber <= $this->lastPageNumber;